crates/uv-resolver/src/marker.rs

```diff
@@ -1,63 +1,84 @@
-use pubgrub::Range;
+use pubgrub::Ranges;
+
 use uv_pep440::Version;
 use uv_pep508::{CanonicalMarkerValueVersion, MarkerTree, MarkerTreeKind};
 
 use crate::requires_python::{LowerBound, RequiresPythonRange, UpperBound};
 
 /// Returns the bounding Python versions that can satisfy the [`MarkerTree`], if it's constrained.
 pub(crate) fn requires_python(tree: MarkerTree) -> Option<RequiresPythonRange> {
-    fn collect_python_markers(tree: MarkerTree, markers: &mut Vec<Range<Version>>) {
+    /// Collect the Python version markers from the tree.
+    ///
+    /// Specifically, performs a DFS to collect all Python requirements on the path to every
+    /// `MarkerTreeKind::True` node.
+    fn collect_python_markers(
+        tree: MarkerTree,
+        markers: &mut Vec<Vec<Ranges<Version>>>,
```

Comment:
The search here only cares about a single type of node (`CanonicalMarkerValueVersion::PythonFullVersion`), and importantly that node can only show up once on a given solution path in the decision diagram. I believe this means that you shouldn't need an inner `Vec` here, just an `Option`, which should simplify this a little bit. There is no list of versions to take an intersection of for a given solution.